enhancements/machine-api/control-plane-machineset-vertical-scaling.md

```diff
@@ -0,0 +1,348 @@
+
+### Workflow Description
+
+1. The OpenShift administrator installs the
+   `control-plane-machine-set-operator`.
+2. The OpenShift adminstrator creates a valid `control-plane-machine-set`
+   `configmap` in the `openshift-machine-api` namespace (or the respective
+   namespace `control-plane-machineset-operator` is running in), to configure
+   the automatic vertical scaling.
+
+### API Extensions
+
+This proposal does not require API extensions.
+
+### Implementation Details/Notes/Constraints [optional]
+
+The implementation for this feature, requires a data source to gather current
+CPU and memory usage for nodes, as well as available resources on the nodes.
+
+To gather available resource information the default
+[Node](https://kubernetes.io/docs/reference/generated/kubernetes-api/v1.27/#node-v1-core)
+API endpoint can be used.
+
+Using the [horizontal pod
+autoscaler](https://kubernetes.io/docs/tasks/run-application/horizontal-pod-autoscale/)
+as a reference, they are using the [metrics
+API](https://kubernetes.io/docs/tasks/debug/debug-cluster/resource-metrics-pipeline/)
+to gather performance statistics for nodes. 
+
+Using the metrics API has the least requirements as it does not expect a running
+Prometheus monitoring stack. Using Prometheus generated data could make the
+implementation easier with respect to data aggregation.
+
+Data aggregation will be required using the metrics API, as decisions about
+scaling nodes up and down, must not be performed based on performance usage at a
+single point in time. Instead the operator will have to gather and store usage
+data for a certain period of time, to decide if scaling is required.
+
+#### Hypershift
+
+Hypershift does not use `ControlPlaneMachineSet` to manage the control-planes
+for hosted clusters, as those are not running on their own virtual machines.
+
+Because of this, no special handling should be required.
+
+### Risks and Mitigations
+
+#### Risk 1: Higher costs when scaling up
+
+Scaling up will increase costs for running the cluster, as bigger nodes will
+incur higher costs from the cloud provider.
+
+#### Risk 2: Cluster instability due to scaling down
+
+Scaling down might also impact the clusters ability to accommodate a workload if
+resource usage was uncharacteristically low for the detection period of scaling.
+This turns into an issue, if the cluster still would have required the bigger
+control plane nodes.
+
+To mitigate this issue the scale down behavior should be disabled by default and
+require an explicit opt-in by the cluster administrator.
+
+#### Risk 3: Control plane node churn
+
+Automatically scaling up and down control planes could impact cluster
+reliability and availability, if performed too often.
+
+As such the additional logic must ensure that there is a grace period between
+changes to the machine size, during which no more scaling will be performed.
+
+### Drawbacks
+
+Some users might not want this feature and rather opt to scale control planes
```

Comment:
It will be opt in right?

Reply:
Yes - in case of OSD we might opt in the customer as SREP, but for plain OCP it would be disabled by default (or in case of no configuration, nothing would happen).